src/mame/machine/tsconf.cpp

```diff
@@ -0,0 +1,622 @@
+// license:BSD-3-Clause
+// copyright-holders:Andrei I. Holub
+
+#include "includes/tsconf.h"
+
+#define W0_RAM (BIT(MEM_CONFIG, 3))
+#define NW0_MAP (BIT(MEM_CONFIG, 2))
+#define W0_WE (BIT(MEM_CONFIG, 1))
+#define ROM128 (BIT(MEM_CONFIG, 0))
+#define RRES (BIT(V_CONFIG, 6, 2))
+
+#define VM static_cast<v_mode>(BIT(V_CONFIG, 0, 2))
+
+static constexpr rectangle resolution_info[5] = {
+	rectangle(52, 256 + 51, 48, 192 + 47), // 52|256|52 x 48-192-48
+	rectangle(20, 320 + 19, 44, 200 + 43), // 20|320|20 x 44-200-44
+	rectangle(20, 320 + 19, 24, 240 + 23), // 20|320|20 x 24-240-24
+	rectangle(0, 360 - 1, 0, 288 - 1),	   //  0|360|0  x  0-288-0
+	rectangle(0, 640 - 1, 0, 240 - 1)	   // text
+};
+
+enum v_mode : u8
+{
+	VM_ZX = 0,
+	VM_16C,
+	VM_256C,
+	VM_TXT
+};
+
+void tsconf_state::tsconf_update_bank1()
+{
+
+	//W0_WE
+	if (NW0_MAP)
+	{
+		m_ROMSelection = PAGE0;
+	}
+	else
+	{
+		/* ROM: 0-SYS, 1-DOS, 2-128, 3-48 */
+		if (m_beta->started() && m_beta->is_active())
+		{
+			m_ROMSelection = ROM128;
+		}
+		else
+		{
+			m_ROMSelection = 0x02 | ROM128;
+		}
+		m_ROMSelection |= (PAGE0 & 0xfc);
+	}
+
+	uint8_t *rom0;
+	if (W0_RAM)
+	{
+		rom0 = m_ram->pointer() + PAGE(m_ROMSelection);
+		m_bank1->set_base(rom0);
+		LOGDEBUG("PAGE0: RAM %X\n", m_ROMSelection);
+	}
+	else
+	{
+		rom0 = &m_p_rom[0x10000 + PAGE(m_ROMSelection & 0x1f)];
+		m_bank1->set_base(rom0);
+		LOGDEBUG("PAGE0: ROM %X\n", m_ROMSelection & 0x1f);
+	}
```

Comment:
Can you do this with configured banks rather than with `set_base`?  It makes save state support a lot simpler.

Reply:
Can you collaborate on this? I copied this approach from base ZX.
Given:
* large ROM space + large RAM space
* CPU only access 4 banks which setup with various handlers
* ROM/RAM pages can be mapped to existing banks

Reply:
Also that's how other spectrum clones behaves at this point. I'm glad to understand how and fix everywhere but possibly later.

Reply:
@cuavas everything addressed except this one. Can you do attempt to review that again?

Reply:
There are two ways you can use the memory banks:
* You can set the base address directly like this.  The advantage is that you get a lot of flexibility.  The disadvantage is saving and loading state won’t be handled correctly – you need to implement a post-load handler to set the correct base address for each memory bank.
* You can configure possible base addresses at start and then select an entry by index.  This is a bit less flexible – you can only do it if there are a finite and reasonable number of base addresses that the bank can point at.  The advantage is that it saves and loads state automatically.

For a simple example of using pre-configured bank base addresses, you can look at xain.cpp:
* Each bank has two possible base addresses, configured on start: https://github.com/mamedev/mame/blob/mame0238/src/mame/drivers/xain.cpp#L464
* Handlers set the bank bases by index: https://github.com/mamedev/mame/blob/mame0238/src/mame/drivers/xain.cpp#L208

Depending on the complexity, you may not be able to use pre-configured banking.  That’s fine if you can’t, it just means a bit more work to add save state support.

Reply:
Thank you for the directions! If you don't mind, promise I return to this a little bit later, because think something like sprites are more important at this point.